easy-ui-react/src/Menu/utilities.ts

```diff
@@ -36,3 +40,55 @@ export function getUnmergedPopoverStyles(
     ),
   };
 }
+
+export function useMenuSelectAll<T>(
+  selectedKeys: Set<Key>,
+  state: TreeState<T>,
+  isFocused: boolean,
+): boolean {
+  // Get collection size minus section item type
+  const itemSize = [...state.collection].reduce((acc, item) => {
+    if (item.type === "section") {
+      acc += [...item.childNodes].length;
+    } else {
+      acc += 1;
+    }
+    return acc;
+  }, 0);
+
+  const isSelectAll = selectedKeys.has(SELECT_ALL_KEY);
+  const selectedSize = selectedKeys.size;
+  const isSelectAllIndeterminate = selectedSize > 0 && !isSelectAll;
+
+  React.useEffect(() => {
+    // This prevent infinite loop
+    if (!state.selectionManager.lastSelectedKey) return;
+
+    if (!isSelectAll && selectedSize === itemSize - 1) {
+      if (isFocused) {
+        // Deselect all checkboxes when Select ALL is deselected
+        state.selectionManager.clearSelection();
+      } else {
+        // Select Select ALL checkbox when all other checkboxes are selected
+        state.selectionManager.selectAll();
+      }
+    } else if (isSelectAll && selectedSize > 0) {
+      if (isFocused) {
+        // Select all checkboxes when Seleclt ALL is selected
+        state.selectionManager.selectAll();
+      } else {
+        // Deselect and mark Select ALL checkbox as indeterminate when deselect one of the checkbox
+        state.selectionManager.select(SELECT_ALL_KEY);
+      }
+    }
+  }, [
+    selectedKeys,
+    isFocused,
+    isSelectAll,
+    selectedSize,
+    itemSize,
+    state.selectionManager,
+  ]);
```

Comment:
**suggestion**: i'm concerned with having the state be modified through a connection with the rendering cycle via `useEffect` and using the menu item's `isFocused`. that paradigm often opens up a component to race conditions and fragility in the future since it's harder to guard against related changes that could affect the behavior.

i would suggest instead for this to be purely event based. this could be done by extending the `onAction` function in `MenuOverlay` and running the selection logic in there based on the SELECT_ALL key. let me know if that makes sense. i can help walk you through the idea.